platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/generic/GenericHtmlTestReportGenerator.java

```diff
@@ -52,53 +63,92 @@
  * <p>
  * The root results are recorded into `index.html`, and then each parent tells its children to generate starting at `{childName}/index.html`.
  */
-public class GenericHtmlTestReport {
+public abstract class GenericHtmlTestReportGenerator implements TestReportGenerator {
 
-    private static final Logger LOG = Logging.getLogger(GenericHtmlTestReport.class);
+    private static final Logger LOG = Logging.getLogger(GenericHtmlTestReportGenerator.class);
 
-    public static String getFilePath(Path path) {
+    public static String getFilePath(org.gradle.util.Path path) {
         String filePath;
         if (path.segmentCount() == 0) {
             filePath = "index.html";
         } else {
-            filePath = String.join("/", path.segments()) + "/index.html";
+            filePath = String.join("/", Iterables.transform(path.segments(), FileUtils::toSafeFileName)) + "/index.html";
         }
         return filePath;
     }
-
     private final BuildOperationRunner buildOperationRunner;
     private final BuildOperationExecutor buildOperationExecutor;
-    private final List<SerializableTestResultStore.OutputReader> outputReaders;
     private final MetadataRendererRegistry metadataRendererRegistry;
+    private final Path reportsDirectory;
 
-    public GenericHtmlTestReport(
+    @Inject
+    public GenericHtmlTestReportGenerator(
         BuildOperationRunner buildOperationRunner,
         BuildOperationExecutor buildOperationExecutor,
-        List<SerializableTestResultStore.OutputReader> outputReaders,
-        MetadataRendererRegistry metadataRendererRegistry
+        MetadataRendererRegistry metadataRendererRegistry,
+        Path reportsDirectory
     ) {
         this.buildOperationRunner = buildOperationRunner;
         this.buildOperationExecutor = buildOperationExecutor;
-        this.outputReaders = outputReaders;
         this.metadataRendererRegistry = metadataRendererRegistry;
+        this.reportsDirectory = reportsDirectory;
+    }
+
+    @Override
+    @Nullable
+    public Path generate(List<Path> resultsDirectories) {
+        List<SerializableTestResultStore> stores = resultsDirectories.stream()
+            .distinct()
+            .map(SerializableTestResultStore::new)
+            .filter(SerializableTestResultStore::hasResults)
+            .collect(Collectors.toList());
+
+        if (stores.stream().noneMatch(SerializableTestResultStore::hasResults)) {
+            return null;
```

Comment:
💭 I wonder if this would be better as an exception?  Why call `generate` at all if there are no results?